### PR TITLE
Provide context for Reactive Rest Client exceptions

### LINKED ISSUE #22090: MP Rest Client - meaningless exceptions
### Describe the bug

I'm using MicroProfile Rest Client with Quarkus and I have an issue with rest client exceptions.

My use case is simple.

If I make mistake in rest client URL I'll get 404 and following stacktrace

```
2021-12-09 21:33:16,635 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: org.jboss.resteasy.reactive.ClientWebApplicationException: Unknown error, status code 404
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.unwrapException(RestClientRequestContext.java:145)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.handleException(AbstractResteasyReactiveContext.java:320)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:164)
        at org.jboss.resteasy.reactive.client.impl.RestClientRequestContext$1.lambda$execute$0(RestClientRequestContext.java:251)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63)
        at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: javax.ws.rs.WebApplicationException: Unknown error, status code 404
        at io.quarkus.rest.client.reactive.runtime.DefaultMicroprofileRestClientExceptionMapper.toThrowable(DefaultMicroprofileRestClientExceptionMapper.java:17)
        at io.quarkus.rest.client.reactive.runtime.MicroProfileRestClientResponseFilter.filter(MicroProfileRestClientResponseFilter.java:36)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:20)
        at org.jboss.resteasy.reactive.client.handlers.ClientResponseFilterRestHandler.handle(ClientResponseFilterRestHandler.java:9)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)
        ... 11 more
```

This stacktrace is totally meaningless.

After reading this stacktrace I don't know:

* Which of rest clients failed  ```MyRestClient```
* What rest client method was called  ```MyRestClient.apiCall() or https://my-api:8080/endpoint```
* Where it was called from ```MyResource.callAPI() -> MySerivice.apiCall() -> MyRestClient.apiCall()```

With ExceptionMapper I can extract rest client endpoint details from ClientWebApplicationException. Unfortunately I can't find any details about the call side.

How can I extract full stacktrace of the call side like "MyResource.callAPI() -> MySerivice.apiCall() -> MyRestClient.apiCall()".
I've tried creating my own ResponseExceptionMapper but I can't find nothing about the call side there.

At the moment only try-catch at the call side can give me full details about rest client call side.

### Expected behavior

Quarkus should provide tdetails about rest client exception in stacktace out of the box - it's rather basic use case

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
We now add the exact method that caused the exception
in order to make it easier for users to identity where
the exception is coming from

Fixes: #22090

### CODE REVIEW

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java [hunk 1]
```diff
@@ -140,8 +143,16 @@ public void abort() {
     protected Throwable unwrapException(Throwable t) {
         var res = super.unwrapException(t);
         if (res instanceof WebApplicationException) {
-            WebApplicationException webApplicationException = (WebApplicationException) res;
-            return new ClientWebApplicationException(webApplicationException.getMessage(), webApplicationException,
+            var webApplicationException = (WebApplicationException) res;
+            var message = webApplicationException.getMessage();
+            var invokedMethodObject = properties.get(MP_INVOKED_METHOD_PROP);
+            if ((invokedMethodObject instanceof Method)) {
```
Comment (michalszynkiewicz) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java:149]: double brackets :)

**Review by michalszynkiewicz — APPROVED** [review]

**Comment by michalszynkiewicz** [comment]
that's a really useful thing, thanks @geoand !

**Comment by michalszynkiewicz** [comment]
I think we coudl backport it, WDYT?

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java (hunk 1 above)
Comment (geoand) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/RestClientRequestContext.java]: Oops! Fixed

**Comment by geoand** [comment]
Yup, very good idea

**Comment by geoand** [comment]
@michalszynkiewicz I had to exclude one of the TCK tests because it was asserting the exact exception message

**Comment by geoand** [comment]
I introduced a property for reverting to the spec behavior simply so we can pass the stupid tests...

**Comment by geoand** [comment]
@radcortez `RestClientOpenTemetry` test seems to be pretty flaky. I've seen it fail a few times the last few days

**Comment by radcortez** [comment]
Yes, I've already noticed it too. I'll have a look.

**Comment by geoand** [comment]
Thanks

**Comment by gwenneg** [comment]
Is there a way to get the old exception message back? I tried `quarkus.rest-client.disable-contextual-error-messages=true` but it's not working from our app 😢

**Comment by gwenneg** [comment]
I'm asking because I need the exception message to be unchanged, otherwise it will break something (which could be fixed with an ugly workaround) in our app.

**Comment by geoand** [comment]
For the time being, `quarkus.rest-client.disable-contextual-error-messages` only works as a system property

**Comment by gwenneg** [comment]
Thanks! I can disable the new behavior with a `System.setProperty` instruction at application startup indeed.

It would be great if `quarkus.rest-client.disable-contextual-error-messages` could become a full-fledge configuration key in a future release.

**Comment by geoand** [comment]
There is https://github.com/quarkusio/quarkus/issues/22777 that tracks exactly that if someone wants to take it up :)

**Comment by gwenneg** [comment]
Isn't https://github.com/quarkusio/quarkus/issues/22777 tracking it as you initially commented?

**Comment by geoand** [comment]
Right, I got mixed up reading too many issues at the same time...

**Comment by TomasHofman** [comment]
Now I'm bit confused by the exchange, but I take it nobody is working on making the `quarkus.rest-client.disable-contextual-error-messages` a configuration property yet. I will try to add it to the config root.

**Comment by geoand** [comment]
Yeah, exactly that @TomasHofman. Sorry for the mixup

**Comment by TomasHofman** [comment]
No problem, thanks for confirming :).

**Pull request merged**